src/main/java/com/google/api/codegen/config/GapicProductConfig.java

```diff
@@ -436,52 +455,183 @@ public static GapicProductConfig createDummyInstance(
 
   private static ImmutableMap<String, ResourceNameConfig> createResourceNameConfigs(
       DiagCollector diagCollector, ConfigProto configProto, TargetLanguage language) {
-    return createResourceNameConfigs(diagCollector, configProto, null, language, null);
+    return createResourceNameConfigs(
+        diagCollector, configProto, null, language, ImmutableMap.of(), ImmutableMap.of(), null);
   }
 
+  /**
+   * Create all the ResourceNameOneofConfig from the protofile and GAPIC config, and let the GAPIC
+   * config resourceNames override the protofile resourceNames in event of clashes.
+   */
   @VisibleForTesting
+  @Nullable
   static ImmutableMap<String, ResourceNameConfig> createResourceNameConfigs(
       DiagCollector diagCollector,
       ConfigProto configProto,
       @Nullable List<ProtoFile> protoFiles,
       TargetLanguage language,
+      Map<Resource, ProtoFile> resourceDefs,
+      Map<ResourceSet, ProtoFile> resourceSetDefs,
       ProtoParser protoParser) {
     ProtoFile file = null;
     if (protoFiles != null) {
       file = protoFiles.get(0);
     }
-    ImmutableMap<String, SingleResourceNameConfig> singleResourceNameConfigs =
-        createSingleResourceNameConfigs(
-            diagCollector, configProto, protoFiles, language, protoParser);
+    ImmutableMap<String, SingleResourceNameConfig> singleResourceNameConfigsFromGapicConfig =
+        createSingleResourceNameConfigs(diagCollector, configProto, protoFiles, language);
     ImmutableMap<String, FixedResourceNameConfig> fixedResourceNameConfigs =
         createFixedResourceNameConfigs(
             diagCollector, configProto.getFixedResourceNameValuesList(), file);
-    ImmutableMap<String, ResourceNameOneofConfig> resourceNameOneofConfigs =
+    ImmutableMap<String, ResourceNameOneofConfig> resourceNameOneofConfigsFromGapicConfig =
         createResourceNameOneofConfigs(
             diagCollector,
             configProto.getCollectionOneofsList(),
-            singleResourceNameConfigs,
+            singleResourceNameConfigsFromGapicConfig,
             fixedResourceNameConfigs,
             file);
+    if (diagCollector.getErrorCount() > 0) {
+      ToolUtil.reportDiags(diagCollector, true);
+      return null;
+    }
+
+    ImmutableMap<String, SingleResourceNameConfig>
+        fullyQualifiedSingleResourceNameConfigsFromProtoFile =
+            createSingleResourceNameConfigsFromProtoFile(diagCollector, resourceDefs, protoParser);
+    ImmutableMap<String, ResourceNameOneofConfig> resourceNameOneofConfigsFromProtoFile =
+        createResourceNameOneofConfigsFromProtoFile(
+            diagCollector,
+            fullyQualifiedSingleResourceNameConfigsFromProtoFile,
+            resourceSetDefs,
+            protoParser);
+
+    // Populate a SingleResourceNameConfigs map, using just the unqualified names.
+    Map<String, SingleResourceNameConfig> singleResourceConfigsFromProtoFile =
+        new LinkedHashMap<>();
```

Comment:
Just use hash map, I assume the `resourceNameConfigs` below will reorder things anyway?

Reply:
ah true